gensim/models/doc2vec.py

```diff
@@ -313,33 +642,88 @@ def note_doctag(self, key, document_no, document_length):
-        candidate_offset = i_index - self.max_rawint - 1
-        if 0 <= candidate_offset < len(self.offset2doctag):
-            return self.offset2doctag[candidate_offset]
+class Doc2VecTrainables(Word2VecTrainables):
+    def __init__(self, mapfile_path=None, dm=1, dm_concat=0, dm_tag_count=1, vector_size=100, seed=1, hs=0, negative=5,
+                 hashfxn=hash, window=5):
+        super(Doc2VecTrainables, self).__init__(
+            vector_size=vector_size, seed=seed, hashfxn=hashfxn)
+        if dm and dm_concat:
+            self.layer1_size = (dm_tag_count + (2 * window)) * self.vector_size
+            logger.info("using concatenative %d-dimensional layer1", self.layer1_size)
+        self.vectors_docs = []
+        self.mapfile_path = mapfile_path
+
+    def reset_weights(self, hs, negative, vocabulary=None):
+        super(Doc2VecTrainables, self).reset_weights(hs, negative, vocabulary=vocabulary)
+        self.reset_doc_weights(vocabulary=vocabulary)
+
+    def reset_doc_weights(self, vocabulary=None):
+        length = max(len(vocabulary.doctags), vocabulary.count)
+        if self.mapfile_path:
+            self.vectors_docs = np_memmap(
+                self.mapfile_path + '.vectors_docs', dtype=REAL, mode='w+', shape=(length, self.vector_size)
+            )
+            self.vectors_docs_lockf = np_memmap(
+                self.mapfile_path + '.vectors_docs_lockf', dtype=REAL, mode='w+', shape=(length,)
+            )
+            self.vectors_docs_lockf.fill(1.0)
         else:
-            return i_index
+            self.vectors_docs = empty((length, self.vector_size), dtype=REAL)
+            self.vectors_docs_lockf = ones((length,), dtype=REAL)  # zeros suppress learning
+
+        for i in xrange(length):
+            # construct deterministic seed from index AND model seed
+            seed = "%d %s" % (
+                self.seed, Doc2VecKeyedVectors._index_to_doctag(i, vocabulary.offset2doctag, vocabulary.max_rawint))
+            self.vectors_docs[i] = self.seeded_vector(seed)
+
+    def save(self, *args, **kwargs):
+        # don't bother storing the cached normalized vectors
+        kwargs['ignore'] = kwargs.get('ignore', ['vectors_norm', 'vectors_docs_norm'])
+        super(Word2VecTrainables, self).save(*args, **kwargs)
+
+    def _get_doctag_trainables(self, doc_words):
+        doctag_vectors = zeros((1, self.vector_size), dtype=REAL)
+        doctag_vectors[0] = self.seeded_vector(' '.join(doc_words))
+        doctag_locks = ones(1, dtype=REAL)
+        return doctag_vectors, doctag_locks
+
+
+class Doc2VecKeyedVectors(BaseKeyedVectors):
+    def __init__(self):
+        self.doctags = {}  # string -> Doctag (only filled if necessary)
+        self.max_rawint = -1  # highest rawint-indexed doctag
+        self.offset2doctag = []  # int offset-past-(max_rawint+1) -> String (only filled if necessary)
+        self.count = 0
+        self.vectors_docs = []
+        self.mapfile_path = None
+
+    @property
```

Comment:
These seem to be for backwards compatibility, right? If yes, best to add `@deprecated` annotations. We decide, in which gensim version the deprecated code will be removed and make that information a part of the deprecation message. @menshikh-iv, thoughts?

Reply:
Agree with you, I already mention it earlier https://github.com/RaRe-Technologies/gensim/pull/1777#discussion_r160505450